File: zone-server/src/main/kotlin/net/bestia/zone/dialog/conversation/InteractCMSG.kt
package net.bestia.zone.dialog.conversation

import net.bestia.bnet.proto.InteractCmsgProto
import net.bestia.zone.message.CMSG
import net.bestia.zone.util.EntityId

/**
 * "I want to deal with that thing."
 *
 * Says nothing about how, because the server decides what interacting with an entity means. Talking is
 * the only answer today, which is why this lives beside the conversation - it moves the day a second
 * kind of interaction exists.
 */
data class InteractCMSG(
  override val playerId: Long,
  val targetEntityId: EntityId,
) : CMSG {

  companion object {
    fun fromBnet(accountId: Long, msg: InteractCmsgProto.InteractCMSG): InteractCMSG {
      return InteractCMSG(playerId = accountId, targetEntityId = msg.entityId)
    }
  }
}

File: bestia-client/src/Bnet/Message/System/InteractCMSG.cs
using Bnet;

namespace BestiaBehemothClient.Bnet.Message.System
{
  /// <summary>
  /// Asks the server to deal with an entity. What that means is the server's decision - talking to
  /// a townsperson today. Answered with a ConversationSMSG, or with nothing at all if the target is
  /// not something this player can interact with.
  /// </summary>
  public partial class InteractCMSG : ICMSG
  {
    public ulong EntityId { get; set; }

    public override Envelope ToEnvelope()
    {
      return new Envelope
      {
        Interact = new global::Bnet.InteractCMSG
        {
          EntityId = EntityId
        }
      };
    }
  }
}
